File: android_app/lib/services/friends_service.dart
import 'package:android_app/generated/l10n.dart';
import 'package:android_app/models/enums/friend_status.dart';
import 'package:android_app/providers/uri_provider.dart';
import 'package:android_app/routes/app_routes.dart';
import 'package:android_app/services/sockets/socket_connection_service.dart';
import 'package:android_app/services/user_info_service.dart';
import 'package:android_app/states/friends_state.dart';
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';
import '../services/auth_service.dart';
import '../models/friend_request.dart';
import 'package:android_app/services/notification_service.dart';
import 'package:audioplayers/audioplayers.dart';

final friendsServiceProvider =
    StateNotifierProvider<FriendsNotifier, FriendsState>((ref) {
  final authService = ref.read(authServiceProvider);
  final socketService = ref.read(socketConnectionServiceProvider);
  final notificationService = ref.read(notificationServiceProvider);
  final baseUrl = ref.read(apiUrlProvider);

  return FriendsNotifier(
    authService: authService,
    socketService: socketService,
    notificationService: notificationService,
    baseUrl: baseUrl,
  );
});

class FriendsNotifier extends StateNotifier<FriendsState> {
  final AuthService _authService;
  final SocketConnectionService _socketService;
  final NotificationService _notificationService;
  final AudioPlayer _audioPlayer = AudioPlayer();
  final String _baseUrl;

  FriendsNotifier({
    required AuthService authService,
    required SocketConnectionService socketService,
    required NotificationService notificationService,
    required String baseUrl,
  })  : _authService = authService,
        _socketService = socketService,
        _notificationService = notificationService,
        _baseUrl = baseUrl,
        super(FriendsState()) {
    authService.onConnectionSubject.subscribe((_) {
      init();
    });

    authService.onDisconnectionSubject.subscribe((_) {
      destroy();
    });
  }

  void init() {
    _socketService.onUnique('friend-socket:receivedRequest', _addReceivedRequest);
    _socketService.onUnique('friend-socket:removedRequest', _removeReceivedRequest);
    _socketService.onUnique('friend-socket:acceptedRequest', _onAcceptedRequest);

    _socketService.onUnique('friend-socket:removedFriend', _friendRemoved);
    _socketService.onUnique('friend-socket:addedFriend', _friendAdded);

    _socketService.onUnique('friend-socket:requestUpdated', (_) {
      refreshReceivedRequest();
      refreshSentRequests();
    });

    _socketService.onUnique('friend-socket:relationUpdated', (_) {
      refreshFriends();
    });

    _socketService.emit('friend-socket:join', _authService.getSessionId());

    refreshFriends();
    refreshSentRequests();
    refreshReceivedRequestsOnLogIn();
  }

  void destroy() {
    _socketService.off('friend-socket:receivedRequest');
    _socketService.off('friend-socket:removedRequest');
    _socketService.off('friend-socket:acceptedRequest');

    _socketService.off('friend-socket:removedFriend');
    _socketService.off('friend-socket:addedFriend');

    _socketService.off('friend-socket:requestUpdated');
    _socketService.off('friend-socket:relationUpdated');
  }

  Future<void> refreshFriends() async {
    final response = await http.get(
      Uri.parse('$_baseUrl/accounts/me/friends'),
      headers: _getSessionIdHeaders(),
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      final List<String> friends = data.map((e) => e.toString()).toList();
      state = state.copyWith(friends: friends);
    }
  }

  Future<void> refreshSentRequests() async {
    final response = await http.get(
      Uri.parse('$_baseUrl/accounts/me/friends/requests/sent'),
      headers: _getSessionIdHeaders(),
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      final List<FriendRequest> requests = data
          .map((e) => FriendRequest.fromJson(e as Map<String, dynamic>))
          .toList();
      state = state.copyWith(sentRequests: requests);
    }
  }

  Future<void> refreshReceivedRequest() async {
    final response = await http.get(
      Uri.parse('$_baseUrl/accounts/me/friends/requests/received'),
      headers: _getSessionIdHeaders(),
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      final List<FriendRequest> requests = data
          .map((e) => FriendRequest.fromJson(e as Map<String, dynamic>))
          .toList();
      state = state.copyWith(receivedRequests: requests);
    }
  }

  Future<void> refreshReceivedRequestsOnLogIn() async {
    final response = await http.get(
      Uri.parse('$_baseUrl/accounts/me/friends/requests/received'),
      headers: _getSessionIdHeaders(),
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      final List<FriendRequest> requests = data
          .map((e) => FriendRequest.fromJson(e as Map<String, dynamic>))
          .toList();
      state = state.copyWith(receivedRequests: requests);

      final unseenRequests =
          requests.where((request) => !request.seen).toList();
      _showUnseenRequestsNotification(unseenRequests);
    }
  }

  Future<void> sendFriendRequest(String username) async {
    await http.post(
      Uri.parse('$_baseUrl/accounts/me/friends/requests/$username'),
      headers: _getSessionIdHeaders(),
    );

    await refreshSentRequests();
  }

  Future<void> deleteFriendRequest(String username) async {
    final response = await http.delete(
      Uri.parse('$_baseUrl/accounts/me/friends/requests/$username'),
      headers: _getSessionIdHeaders(),
    );

    if (response.statusCode == 200) {
      await refreshSentRequests();
    }
  }

  Future<void> answerFriendRequestByUsername(String username, bool answer) {
    var request = state.receivedRequests
        .firstWhere((request) => request.senderUsername == username);
    return answerFriendRequest(request.id, answer);
  }

  Future<void> answerFriendRequest(String requestId, bool answer) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/accounts/me/friends/requests/answer/$requestId'),
      headers: _getSessionIdHeaders(),
      body: json.encode({'answer': answer}),
    );

    if (response.statusCode == 200) {
      await refreshReceivedRequest();
    }
  }

  void sawFriendRequest(String requestId) {
    _socketService.emit('friend-socket:seenRequest', requestId);
  }

  Future<void> removeFriend(String username) async {
    final response = await http.delete(
      Uri.parse('$_baseUrl/accounts/me/friends/$username'),
      headers: _getSessionIdHeaders(),
    );

    if (response.statusCode == 200) {
      await refreshFriends();
    }
  }

  void _friendAdded(dynamic data) {
    final username = data.toString();
    final currentFriends = List<String>.from(state.friends);
    state = state.copyWith(friends: [...currentFriends, username]);
  }

  void _friendRemoved(dynamic data) {
    final username = data.toString();
    final currentFriends = List<String>.from(state.friends);
    state = state.copyWith(
      friends: currentFriends.where((friend) => friend != username).toList(),
    );
  }

  void _addReceivedRequest(dynamic data) {
    final request = FriendRequest.fromJson(data as Map<String, dynamic>);
    final currentRequests = List<FriendRequest>.from(state.receivedRequests);
    state = state.copyWith(receivedRequests: [...currentRequests, request]);

    _audioPlayer.play(AssetSource('sounds/message1.mp3'));

    _showFriendRequestNotification(request);
    sawFriendRequest(request.id);
  }

  void _removeReceivedRequest(dynamic data) {
    final removedRequest = FriendRequest.fromJson(data as Map<String, dynamic>);
    final currentRequests = List<FriendRequest>.from(state.receivedRequests);
    state = state.copyWith(
      receivedRequests: currentRequests
          .where((request) =>
              request.senderUsername != removedRequest.senderUsername)
          .toList(),
    );
  }

  void _onAcceptedRequest(dynamic data) {
    final response = data as Map<String, dynamic>;
    final accepterUsername = response['accepterUsername'] as String;
    final answer = response['answer'] as bool;

    final currentRequests = List<FriendRequest>.from(state.sentRequests);
    state = state.copyWith(
      sentRequests: currentRequests
          .where((request) => request.receiverUsername != accepterUsername)
          .toList(),
    );

    if (!answer) return;
    var message = S.current.friend_requests_request_accepted(accepterUsername);

    _audioPlayer.play(AssetSource('sounds/message1.mp3'));
    
    _notificationService.showTopRightNotification(
      message: message,
      icon: Icons.how_to_reg,
    );
  }

  void _showFriendRequestNotification(FriendRequest request) {
    var message =
        S.current.friend_requests_new_request_from(request.senderUsername);
    _notificationService.showTopRightNotification(
      message: message,
      icon: Icons.person_add,
    );
  }

  void _showUnseenRequestsNotification(List<FriendRequest> unseenRequests) {
    if (unseenRequests.isEmpty) return;

    var message =
        S.current.friend_requests_multiple_requests(unseenRequests.length);

    _notificationService.showTopRightNotification(
      message: message,
      icon: Icons.group_add,
      actionCallback: () {
        navigatorKey.currentContext?.go('/personal-profile?tab=friends');
      },
    );

    for (final request in unseenRequests) {
      sawFriendRequest(request.id);
    }
  }

  Map<String, String> _getSessionIdHeaders() {
    return {
      'Content-Type': 'application/json',
      'x-session-id': _authService.getSessionId(),
    };
  }
}

final getFriendStatusProvider =
    AutoDisposeProvider.family<FriendStatus, String>((ref, username) {
  if (username ==
      ref.watch(userInfoServiceProvider.select((state) => state.username))) {
    return FriendStatus.self;
  }

  final friendsSet = {
    ...ref.watch(friendsServiceProvider.select((state) => state.friends))
  };
  if (friendsSet.contains(username)) return FriendStatus.friend;

  final sentRequest =
      ref.watch(friendsServiceProvider.select((state) => state.sentRequests));
  final sentRequestsSet = {
    ...sentRequest.map((request) => request.receiverUsername)
  };
  if (sentRequestsSet.contains(username)) return FriendStatus.sentRequest;

  final receivedRequest = ref
      .watch(friendsServiceProvider.select((state) => state.receivedRequests));
  final receivedRequestsSet = {
    ...receivedRequest.map((request) => request.senderUsername)
  };
  if (receivedRequestsSet.contains(username)) {
    return FriendStatus.receivedRequest;
  }

  return FriendStatus.none;
});

File: client/src/app/services/friends.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable, BehaviorSubject, of, combineLatest, map } from 'rxjs';
import { environment } from 'src/environments/environment';
import { UserAuthenticationService } from '@app/services/user-authentication.service';
import { FriendRequest } from '@app/interfaces/friend-request';
import { SocketFactoryService } from './socket-service/socket-factory.service';
import { SocketService } from './socket-service/socket.service';
import { TranslateService } from '@ngx-translate/core';
import { DialogModalService } from './dialog-modal.service';
import { NotificationComponent } from '@app/components/notification/notification.component';
import { PrivateProfileService } from './private-profile.service';
import { FriendStatus } from '@app/enums/friend-status.enum';
import { UserInfoService } from './user-info.service';
import { ChatsService } from './chat-services/chats-service';
import { AudioService } from '@app/services/audio.service';
import { NOTIFICATION_SOUND_MP3 } from '@app/consts/file-consts';

@Injectable({
    providedIn: 'root',
})
export class FriendsService {
    private friendsSubject = new BehaviorSubject<string[]>([]);
    private sentRequestsSubject = new BehaviorSubject<FriendRequest[]>([]);
    private receivedRequestsSubject = new BehaviorSubject<FriendRequest[]>([]);

    private readonly baseUrl: string = environment.serverUrl + '/accounts';
    private readonly socketService: SocketService;

    private readonly boundHandlers = {
        receivedRequest: this.addReceivedRequest.bind(this),
        removedRequest: this.removeReceivedRequest.bind(this),
        acceptedRequest: this.onAcceptedRequest.bind(this),
        removedFriend: this.friendRemoved.bind(this),
        addedFriend: this.friendAdded.bind(this),
        requestUpdated: this.requestUpdated.bind(this),
        relationUpdated: this.refreshFriends.bind(this)
    };

    get friendsObservable() { return this.friendsSubject.asObservable(); }
    get sentRequestsObservable() { return this.sentRequestsSubject.asObservable(); }
    get receivedRequestsObservable() { return this.receivedRequestsSubject.asObservable(); }

    constructor(
        private readonly http: HttpClient,
        private readonly authService: UserAuthenticationService,
        private readonly userInfoService: UserInfoService,
        private readonly dialogModalService: DialogModalService,
        private readonly translate: TranslateService,
        private readonly privateProfile: PrivateProfileService,
        private readonly chatsService: ChatsService,
        private readonly audioService: AudioService,
        socketFactory: SocketFactoryService,
    ) {
        this.socketService = socketFactory.getSocket();

        this.authService.connectionEvent.subscribe((_) => this.init());
        this.authService.disconnectEvent.subscribe(() => this.destroy());
    }

    getFriends(): Observable<string[]> {
        return this.http.get<string[]>(
            `${this.baseUrl}/me/friends`,
            { headers: this.authService.getSessionIdHeaders() }
        );
    }

    getUserFriendStatus(username: string) {
        if (username == this.userInfoService.user.username)
            return of(FriendStatus.Self);
        return combineLatest([
            this.friendsObservable,
            this.sentRequestsObservable,
            this.receivedRequestsObservable,
        ]).pipe(map(
            ([friends, sent, received]) => this.filterUserStatus(username, friends, sent, received)
        ));
    }

    sendFriendRequest(username: string) {
        this.http.post(
            `${this.baseUrl}/me/friends/requests/${username}`, {},
            { headers: this.authService.getSessionIdHeaders() }
        ).subscribe(() => this.refreshSentRequests());
    }

    deleteFriendRequest(username: string) {
        this.http.delete(
            `${this.baseUrl}/me/friends/requests/${username}`,
            { headers: this.authService.getSessionIdHeaders() }
        ).subscribe(() => this.refreshSentRequests());
    }

    answerFriendRequest(requestId: string, answer: boolean) {
        this.http.post(
            `${this.baseUrl}/me/friends/requests/answer/${requestId}`, { answer },
            { headers: this.authService.getSessionIdHeaders() }
        ).subscribe(() => this.refreshReceivedRequest());
    }

    answerFriendRequestByUsername(username: string, answer: boolean) {
        let request = this.receivedRequestsSubject.getValue().find((request) => request.senderUsername === username)
        if (!request) throw new Error('invalid username');

        this.answerFriendRequest(request.id, answer);
    }

    sawFriendRequest(requestId: string) {
        this.socketService.emit('friend-socket:seenRequest', requestId);
    }

    removeFriend(username: string) {
        return this.http.delete(
            `${this.baseUrl}/me/friends/${username}`, { headers: this.authService.getSessionIdHeaders() }
        ).subscribe(() => this.refreshFriends());
    }

    private init() {
        this.socketService.on('friend-socket:receivedRequest', this.boundHandlers.receivedRequest);
        this.socketService.on('friend-socket:removedRequest', this.boundHandlers.removedRequest);
        this.socketService.on('friend-socket:acceptedRequest', this.boundHandlers.acceptedRequest);
        this.socketService.on('friend-socket:removedFriend', this.boundHandlers.removedFriend);
        this.socketService.on('friend-socket:addedFriend', this.boundHandlers.addedFriend);
        this.socketService.on('friend-socket:requestUpdated', this.boundHandlers.requestUpdated);
        this.socketService.on('friend-socket:relationUpdated', this.boundHandlers.relationUpdated);

        this.socketService.emit('friend-socket:join', this.authService.getSessionId());

        this.refreshFriends();
        this.refreshSentRequests();
        this.refreshReceivedRequestsOnLogIn();
    }

    private destroy() {
        this.socketService.off('friend-socket:receivedRequest', this.boundHandlers.receivedRequest);
        this.socketService.off('friend-socket:removedRequest', this.boundHandlers.removedRequest);
        this.socketService.off('friend-socket:acceptedRequest', this.boundHandlers.acceptedRequest);
        this.socketService.off('friend-socket:removedFriend', this.boundHandlers.removedFriend);
        this.socketService.off('friend-socket:addedFriend', this.boundHandlers.addedFriend);
        this.socketService.off('friend-socket:requestUpdated', this.boundHandlers.requestUpdated);
        this.socketService.off('friend-socket:relationUpdated', this.boundHandlers.relationUpdated);
    }

    private refreshFriends() {
        this.getFriends().subscribe((friends) => this.friendsSubject.next(friends));
    }

    private requestUpdated() {
        this.refreshReceivedRequest();
        this.refreshSentRequests();
    }

    private getSentFriendRequests(): Observable<FriendRequest[]> {
        return this.http.get<FriendRequest[]>(
            `${this.baseUrl}/me/friends/requests/sent`,
            { headers: this.authService.getSessionIdHeaders() }
        );
    }

    private getReceivedFriendRequests(): Observable<FriendRequest[]> {
        return this.http.get<FriendRequest[]>(
            `${this.baseUrl}/me/friends/requests/received`,
            { headers: this.authService.getSessionIdHeaders() }
        );
    }

    private refreshSentRequests() {
        this.getSentFriendRequests().subscribe((requests) => this.sentRequestsSubject.next(requests));
    }

    private refreshReceivedRequest() {
        this.getReceivedFriendRequests().subscribe((requests) => this.receivedRequestsSubject.next(requests));
    }

    private refreshReceivedRequestsOnLogIn() {
        this.getReceivedFriendRequests().subscribe((requests) => {
            this.receivedRequestsSubject.next(requests);
            const unseenRequests = requests.filter((request) => !request.seen);
            this.showUnseenRequestsNotification(unseenRequests);
        });
    }

    private friendAdded(username: string) {
        const currentFriends = this.friendsSubject.value;
        this.friendsSubject.next([...currentFriends, username]);
    }

    private friendRemoved(username: string) {
        const currentFriends = this.friendsSubject.value;
        this.friendsSubject.next(currentFriends.filter(friendUsername => friendUsername !== username));
        this.chatsService.cleanupFriendOnlyChats(username);
    }

    private addReceivedRequest(request: FriendRequest) {
        const currentRequests = this.receivedRequestsSubject.value;
        this.receivedRequestsSubject.next([...currentRequests, request]);

        this.audioService.playSound(NOTIFICATION_SOUND_MP3);
        
        this.showFriendRequestNotification(request);
        this.sawFriendRequest(request.id);
    }

    private removeReceivedRequest(removedRequest: FriendRequest) {
        const currentRequests = this.receivedRequestsSubject.value;
        this.receivedRequestsSubject.next(currentRequests.filter(request => request.senderUsername !== removedRequest.senderUsername));
    }

    private onAcceptedRequest(response: { accepterUsername: string, answer: boolean }) {
        const currentRequests = this.sentRequestsSubject.value;
        this.sentRequestsSubject.next(currentRequests.filter(request => request.receiverUsername !== response.accepterUsername));

        if (!response.answer) return;

        const message = this.translate.instant('FriendRequests.RequestAccepted', { username: response.accepterUsername });
        const data = { message, icon: 'how_to_reg' };
        
        this.audioService.playSound(NOTIFICATION_SOUND_MP3);

        this.dialogModalService.openNotification(NotificationComponent, data);
    }

    private filterUserStatus(username: string, friends: string[], sent: FriendRequest[], received: FriendRequest[]) {
        let friendsSet = new Set(friends);
        if (friendsSet.has(username)) return FriendStatus.Friend;

        let sentSet = new Set(sent.map((sent) => sent.receiverUsername));
        if (sentSet.has(username)) return FriendStatus.SentRequest;

        let receivedSet = new Set(received.map((received) => received.senderUsername));

        if (receivedSet.has(username)) return FriendStatus.ReceivedRequest;
        return FriendStatus.None;
    }

    private showFriendRequestNotification(request: FriendRequest) {
        const message = this.translate.instant('FriendRequests.NewRequestFrom', { username: request.senderUsername });

        const data = { message, icon: 'person_add' };
        this.dialogModalService.openNotification(NotificationComponent, data);
    }

    private showUnseenRequestsNotification(unseenRequests: FriendRequest[]) {
        if (!unseenRequests.length) return;

        const data = {
            message: this.translate.instant('FriendRequests.MultipleRequests', { count: unseenRequests.length }),
            icon: 'group_add',
            arrowCallback: () => this.privateProfile.goToPage('friends'),
        };

        this.dialogModalService.openNotification(NotificationComponent, data);
        unseenRequests.forEach((request) => this.sawFriendRequest(request.id));
    }
}
